doc/eventcounter.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH EVENTCOUNTER 1 "March  5, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
eventcounter, eventgenerator, eventlogger, eventselect, eventsink,
eventsource, viewevents \- utility applications which can be used
together with other MUSIC-aware applications in multi-simulations
.SH SYNOPSIS
.B eventcounter
.RB [ "\-tbmh " ]
.I "n_units prefix "
[
.I suffix
]
.br
.B eventgenerator
.RB [ "\-tbfmih " ]
.I "n_units"
.br
.B eventlogger
.RB [ "\-tlbmih " ]
.br
.B eventselect
.RB [ "\-th " ]
.I n_units units
.br
.B eventsink
.RB [ "\-tmih " ]
.I "n_units prefix "
[
.I suffix
]
.br
.B eventsource
.RB [ "\-tbmih " ]
.I "n_units prefix "
[
.I suffix
]
.br
.B viewevents
.RB [ "\-tshT " ]
.I "configfile"
.SH DESCRIPTION
This manual page documents briefly a set of MUSIC-aware utility
applications provided together with MUSIC library.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBeventcounter\fP receives spikes through a MUSIC input port, counts
all spikes for each index and writes the frequencies to a set of files
with names
.I "prefix rank suffix"
.PP
\fBeventgenerator\fP generates spikes from a Poisson distribution.
.PP
\fBeventlogger\fP logs spikes from a MUSIC port.
.PP
\fBeventselect\fP receives events from an input port of width
.I "n_units "
and sends events for the subset of id:s specified in the file
.I "units"
.PP
\fBeventsink\fP receives spikes through a MUSIC input port and
writes these to a set of files with names
.I "prefix rank suffix"
.PP
\fBeventsource\fP reads spikes from a set of files with names
.I "prefix rank suffix"
and propagates these spikes through a MUSIC output port.
.PP
\fBviewevents\fP reads spikes from a MUSIC input port and displays
them as a 3D graphical representation.
.SH OPTIONS
The utilities follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-b ticks, \-\-maxbuffered ticks
Specify maximal amount of data buffered.
.TP
.B \-f freq, \-\-frequency
Specify average frequency (default 10 Hz).
.TP
.B \-h, \-\-help
Show usage information.
.TP
.B \-i, \-\-indextype
Select global (default) or local indices. (Used for testing and
benchmarking purposes.)
.TP
.B \-l latency, \-\-acclatency latency
Specify acceptable data latency (s).
.TP
.B \-m type, \-\-imaptype type
Select linear (default) or roundrobin index map.
.TP
.B \-s scaling, \-\-scaletime scaling
Specify real time to simulated time scale factor (s).  If omitted, the
visualization runs at full speed.
.TP
.B \-t timestep, \-\-timestep timestep
Specify time between tick() calls (default 0.01 s).
.TP
.B \-T title, \-\-title title
Specify window title.
.SH SEE ALSO
.BR music (1)
.\" .BR baz (1).
.\" .br
.\" The programs are documented fully by
.\" .IR "The Rise and Fall of a Fooish Bar" ,
.\" available via the Info system.
.SH AUTHOR
MUSIC was written by Mikael Djurfeldt and Örjan Ekeberg for INCF.
viewevents was written by Johannes Hjorth for INCF.
.PP
This manual page was written by Mikael Djurfeldt <mdj@debian.org>,
for the Debian project (but may be used by others).


NEWS
MUSIC NEWS --- history of user-visible changes.
Copyright (C) 2009, 2010 INCF

Please send MUSIC bug reports to music-bugs@incf.org.


Changes in 1.0.7:

* music library

** Bug fix release.


Changes in 1.0.6:

* music library

** Fixes for GCC version 4.4.

** Don't include RPATH in libmusic.so.


Changes in 1.0.5:

* music library

** Optimizations of event routing.  MUSIC now discovers ranges of
   indices which can be handled collectively by one entry in the event
   router, even in the case that this is not apparent from index maps
   but due to the effect of negotiations with the port connected on
   the other side.

** Bug fixes.
   

Changes in 1.0.4:

* test directory

** New example: messages.music

* music library

** Message ports now work correctly.

** Optimizations.


Changes in 1.0.3:

* music library

** Fixes a bug which occurs in Runtime::finalize () when using the
   threading version of the Setup constructor under OpenMPI 1.2.7.

** Other smaller fixes.


Changes in 1.0.2:

* music library

** Better error reporting.

** Bug fixes.

** New README file in test directory.

** New demo in test directory.


Changes in 1.0.1:

* music utility

** Bug fixes

** Improved error reporting

** New option '-m' prints application rank mapping

** New option '-v' prints version information

** Debian packaging support


First release: 1.0

Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


utils/eventcounter.cc
/*
 *  This file is part of MUSIC.
 *  Copyright (C) 2008, 2009 INCF
 *
 *  MUSIC is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  MUSIC is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// Leave as first include---required by BG/L
#include <mpi.h>

#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <cstdlib>
#include <vector>

extern "C" {
#include <unistd.h>
#include <getopt.h>
}

#include <music.hh>

const double DEFAULT_TIMESTEP = 1e-2;

void
usage (int rank)
{
  if (rank == 0)
    {
      std::cerr << "Usage: eventcounter [OPTION...] N_UNITS PREFIX [SUFFIX]" << std::endl
		<< "`eventcounter' receives spikes through a MUSIC input port, counts" << std::endl
		<< "all spikes for each index and writes the frequencies to a set of files" << std::endl
		<< "with names PREFIX RANK SUFFIX" << std::endl << std:: endl
		<< "  -t, --timestep TIMESTEP time between tick() calls (default " << DEFAULT_TIMESTEP << " s)" << std::endl
		<< "  -b, --maxbuffered TICKS maximal amount of data buffered" << std::endl
		<< "  -m, --imaptype TYPE     linear (default) or roundrobin" << std::endl
		<< "  -h, --help              print this help message" << std::endl << std::endl
		<< "Report bugs to <music-bugs@incf.org>." << std::endl;
    }
  exit (1);
}

std::vector<int> counters;

class MyEventHandlerLocal: public MUSIC::EventHandlerLocalIndex {
public:
  void operator () (double t, MUSIC::LocalIndex id)
  {
    ++counters[id];
  }
};

int nUnits;
double timestep = DEFAULT_TIMESTEP;
int    maxbuffered = 0;
string imaptype = "linear";
string prefix;
string suffix = ".dat";

void
getargs (int rank, int argc, char* argv[])
{
  opterr = 0; // handle errors ourselves
  while (1)
    {
      static struct option longOptions[] =
	{
	  {"timestep",  required_argument, 0, 't'},
	  {"maxbuffered", required_argument, 0, 'b'},
	  {"imaptype",  required_argument, 0, 'm'},
	  {"help",      no_argument,       0, 'h'},
	  {0, 0, 0, 0}
	};
      /* `getopt_long' stores the option index here. */
      int option_index = 0;

      // the + below tells getopt_long not to reorder argv
      int c = getopt_long (argc, argv, "+t:b:m:h", longOptions, &option_index);

      /* detect the end of the options */
      if (c == -1)
	break;

      switch (c)
	{
	case 't':
	  timestep = atof (optarg); // NOTE: could do error checking
	  continue;
	case 'b':
	  maxbuffered = atoi (optarg);
	  continue;
	case 'm':
	  imaptype = optarg;
	  if (imaptype != "linear" && imaptype != "roundrobin")
	    {
	      usage (rank);
	      abort ();
	    }
	  continue;
	case '?':
	  break; // ignore unknown options
	case 'h':
	  usage (rank);

	default:
	  abort ();
	}
    }

  if (argc < optind + 2 || argc > optind + 3)
    usage (rank);

  nUnits = atoi (argv[optind]);
  prefix = argv[optind + 1];
  if (argc == optind + 3)
    suffix = argv[optind + 2];
}

int
main (int argc, char *argv[])
{
  MUSIC::Setup* setup = new MUSIC::Setup (argc, argv);
  
  MPI::Intracomm comm = setup->communicator ();
  int nProcesses = comm.Get_size ();
  int rank = comm.Get_rank ();
  
  getargs (rank, argc, argv);

  MUSIC::EventInputPort* in = setup->publishEventInput ("in");
  if (!in->isConnected ())
    {
      if (rank == 0)
	std::cerr << "eventcounter port is not connected" << std::endl;
      comm.Abort (1);
    }

  std::ostringstream spikefile;
  spikefile << prefix << rank << suffix;
  std::ofstream out (spikefile.str ().c_str ());
  if (!out)
    {
      std::cerr << "eventcounter: could not open "
		<< spikefile.str () << " for writing" << std::endl;
      abort ();      
    }

  MyEventHandlerLocal evhandlerLocal;
  
  if (imaptype == "linear")
    {
      int nUnitsPerProcess = nUnits / nProcesses;
      int nLocalUnits = nUnitsPerProcess;
      int rest = nUnits % nProcesses;
      int firstId = nUnitsPerProcess * rank;
      if (rank < rest)
	{
	  firstId += rank;
	  nLocalUnits += 1;
	}
      else
	firstId += rest;
      MUSIC::LinearIndex indices (firstId, nLocalUnits);

      if (maxbuffered > 0)
	in->map (&indices, &evhandlerLocal, 0.0, maxbuffered);
      else
	in->map (&indices, &evhandlerLocal, 0.0);
    }
  else
    {
      std::vector<MUSIC::GlobalIndex> v;
      for (int i = rank; i < nUnits; i += nProcesses)
	v.push_back (i);
      MUSIC::PermutationIndex indices (&v.front (), v.size ());

      if (maxbuffered > 0)
	in->map (&indices, &evhandlerLocal, 0.0, maxbuffered);
      else
	in->map (&indices, &evhandlerLocal, 0.0);
    }

  counters.resize (nUnits);
  for (int i = 0; i < nUnits; ++i)
    counters[i] = 0;

  double stoptime;
  setup->config ("stoptime", &stoptime);

  MUSIC::Runtime* runtime = new MUSIC::Runtime (setup, timestep);

  double time = runtime->time ();
  while (time < stoptime)
    {
      // Retrieve data from other program
      runtime->tick ();
      
      time = runtime->time ();
    }
  for (std::vector<int>::iterator i = counters.begin ();
       i != counters.end ();
       ++i)
    out << static_cast<double> (*i) / stoptime << std::endl;
  out.close ();

  runtime->finalize ();

  delete runtime;

  return 0;
}
